### PR TITLE
rootless: btrfs: annotate error with human-readable hint string

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**- What I did**

Add hints for `Failed to destroy btrfs snapshot <DIR> for <ID>: operation not permitted` error on rootless

Related to issue #41762

**- How I did it**

Wrap the error when running in a user namespace.

**- How to verify it**

```console
$ dockerd-rootless.sh --data-root /mnt/btrfs/varlibdocker
```

```console
$ docker --context=rootless run --rm hello-world
Hello from Docker!
...
ERRO[0006] Error waiting for container: container 59753471c3be3525b46678702e52445aa596da01ccd19e502fde9ff14ec1cfe9: driver "btrfs" failed to remove root filesystem: failed to delete subvolume without root (hint: remount btrfs on "user_subvol_rm_allowed" option, or change the storage driver to "fuse-overlayfs"): Failed to destroy btrfs snapshot /mnt/btrfs/varlibdocker/btrfs/subvolumes for 8a09ee12fd5cb205dfd554a189a389645bb38184703874cd92745690868d845a: operation not permitted 
```

**- Description for the changelog**

rootless: btrfs: annotate error with human-readable hint string

**- A picture of a cute animal (not mandatory but encouraged)**

:penguin:

### CODE REVIEW

File: daemon/graphdriver/btrfs/btrfs.go [hunk 1]
```diff
@@ -633,6 +634,10 @@ func (d *Driver) Remove(id string) error {
 	d.updateQuotaStatus()
 
 	if err := subvolDelete(d.subvolumesDir(), id, d.quotaEnabled); err != nil {
+		if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
+			// https://github.com/moby/moby/issues/41762
+			err = errors.Wrap(err, "failed to delete subvolume without root (hint: remount btrfs on \"user_subvol_rm_allowed\" option, or change the storage driver to \"fuse-overlayfs\")")
+		}
```
Comment (AkihiroSuda) [inline daemon/graphdriver/btrfs/btrfs.go:640]: Note: proper fix is being implemented in https://github.com/moby/moby/pull/42203 , but requires kernel >= 4.18.
So I think we can have both PRs.

File: daemon/graphdriver/btrfs/btrfs.go [hunk 2]
```diff
@@ -633,6 +634,10 @@ func (d *Driver) Remove(id string) error {
 	d.updateQuotaStatus()
 
 	if err := subvolDelete(d.subvolumesDir(), id, d.quotaEnabled); err != nil {
+		if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
```
Comment (thaJeztah) [inline daemon/graphdriver/btrfs/btrfs.go:637]: Would we be able to get an error-type instead of string matching?

**Review by tonistiigi — APPROVED** [review]
I don't think this should be cherry-picked

File: daemon/graphdriver/btrfs/btrfs.go (hunk 2 above)
Comment (AkihiroSuda) [reply daemon/graphdriver/btrfs/btrfs.go:637]: Doesn't seem straightforward https://github.com/moby/moby/blob/8e29dd734d20ddf503fb974190fac42cca71aa42/daemon/graphdriver/btrfs/btrfs.go#L267

**Comment by AkihiroSuda** [comment]
rebased

File: daemon/graphdriver/btrfs/btrfs.go (hunk 2 above)
Comment (thaJeztah) [reply daemon/graphdriver/btrfs/btrfs.go]: Isn't it coming from https://github.com/keybase/client/blob/master/go/vendor/golang.org/x/sys/unix/zerrors_linux_amd64.go#L2921 ?

Was thinking something like `errors.Is(err, os.ErrPermission)`

File: daemon/graphdriver/btrfs/btrfs.go [hunk 3]
```diff
@@ -634,6 +635,9 @@ func (d *Driver) Remove(id string) error {
 
 	if err := subvolDelete(d.subvolumesDir(), id, d.quotaEnabled); err != nil {
 		if d.quotaEnabled {
+			if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
+				err = errors.Wrap(err, "failed to delete subvolume without root (hint: remount btrfs on \"user_subvol_rm_allowed\" option, or update the kernel to >= 4.18, or change the storage driver to \"fuse-overlayfs\")")
```
Comment (thaJeztah) [inline daemon/graphdriver/btrfs/btrfs.go:639]: Perhaps use backticks for this one;

```suggestion
				err = errors.Wrap(err, `failed to delete subvolume without root (hint: remount btrfs on "user_subvol_rm_allowed" option, or update the kernel to >= 4.18, or change the storage driver to "fuse-overlayfs")`)
```

File: daemon/graphdriver/btrfs/btrfs.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ import (
 	"sync"
 	"unsafe"
 
+	"github.com/containerd/containerd/sys"
```
Comment (thaJeztah) [inline daemon/graphdriver/btrfs/btrfs.go:29]: Could you do another rebase, and update this to use `"github.com/containerd/containerd/pkg/userns"` ?

```suggestion
	"github.com/containerd/containerd/pkg/userns"
```


Full diff (with my other suggestion);

```patch
diff --git a/daemon/graphdriver/btrfs/btrfs.go b/daemon/graphdriver/btrfs/btrfs.go
index 677dc7aed4..0d0ee3a207 100644
--- a/daemon/graphdriver/btrfs/btrfs.go
+++ b/daemon/graphdriver/btrfs/btrfs.go
@@ -26,7 +26,7 @@ import (
        "sync"
        "unsafe"

-       "github.com/containerd/containerd/sys"
+       "github.com/containerd/containerd/pkg/userns"
        "github.com/docker/docker/daemon/graphdriver"
        "github.com/docker/docker/pkg/containerfs"
        "github.com/docker/docker/pkg/idtools"
@@ -601,8 +601,8 @@ func (d *Driver) Remove(id string) error {

        if err := subvolDelete(d.subvolumesDir(), id, d.quotaEnabled); err != nil {
                if d.quotaEnabled {
-                       if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
-                               err = errors.Wrap(err, "failed to delete subvolume without root (hint: remount btrfs on \"user_subvol_rm_allowed\" option, or update the kernel to >= 4.18, or change the storage driver to \"fuse-overlayfs\")")
+                       if userns.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
+                               err = errors.Wrap(err, `failed to delete subvolume without root (hint: remount btrfs on "user_subvol_rm_allowed" option, or update the kernel to >= 4.18, or change the storage driver to "fuse-overlayfs")`)
                        }
                        return err
                }
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: daemon/graphdriver/btrfs/btrfs.go (hunk 2 above)
Comment (AkihiroSuda) [reply daemon/graphdriver/btrfs/btrfs.go]: That requires replacing `%v` with `%w`

File: daemon/graphdriver/btrfs/btrfs.go (hunk 3 above)
Comment (AkihiroSuda) [reply daemon/graphdriver/btrfs/btrfs.go]: done

File: daemon/graphdriver/btrfs/btrfs.go (hunk 4 above)
Comment (AkihiroSuda) [reply daemon/graphdriver/btrfs/btrfs.go]: done

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Failure is unrelated; merging

**Pull request merged**